src/operations/query.ts

```diff
@@ -152,42 +149,61 @@ const readSelection = (
         }
       } else {
         const prevData = data[fieldAlias] as Data;
-        data[fieldAlias] = readField(ctx, link, fieldSelect, prevData);
+        data[fieldAlias] = resolveLink(ctx, link, fieldSelect, prevData);
       }
     }
   });
 
   return data;
 };
 
-const readResolverSelection = (
+const resolveResolverResult = (
   ctx: Context,
-  entity: null | Entity | NullArray<Entity>,
+  result: DataField,
   key: string,
   select: SelectionSet,
   prevData: void | Data | Data[]
 ) => {
   // When we are dealing with a list we have to call this method again.
-  if (Array.isArray(entity)) {
+  if (Array.isArray(result)) {
     // @ts-ignore: Link cannot be expressed as a recursive type
-    return entity.map((childEntity, index) => {
+    return result.map((childResult, index) => {
       const data = prevData !== undefined ? prevData[index] : undefined;
       const indexKey = joinKeys(key, `${index}`);
-      return readResolverSelection(ctx, childEntity, indexKey, select, data);
+      return resolveResolverResult(ctx, childResult, indexKey, select, data);
     });
-  } else if (entity === null) {
+  } else if (result === null) {
     return null;
-  } else {
-    const data = prevData === undefined ? Object.create(null) : prevData;
-    const entityKey = keyOfEntity(entity);
-    const childKey = entityKey !== null ? entityKey : key;
+  } else if (isDataOrKey(result)) {
     // We don't need to read the entity after exiting a resolver
     // we can just go on and read the selection further.
-    return readSelection(ctx, childKey, select, data);
+    const data = prevData === undefined ? Object.create(null) : prevData;
+    const childKey =
+      (typeof result === 'string' ? result : keyOfEntity(result)) || key;
+    const selectionResult = readSelection(ctx, childKey, select, data);
+
+    if (selectionResult !== null && typeof result === 'object') {
+      for (key in result) {
+        if (key !== '__typename' && result.hasOwnProperty(key)) {
+          selectionResult[key] = result[key];
+        }
+      }
+    }
+
+    return selectionResult;
+  }
+
+  if (process.env.NODE_ENV !== 'production') {
+    console.warn(
+      'Expected to receive a Link or an Entity from a Resolver but got a Scalar.'
```

Comment:
We might have to include the field + entity here to give some feedback

Reply:
Let's take care of tests in a big "testing PR" and improve warnings there since we're missing a lot of warnings anyway 😅